File: adapters/go/dna_generator.go
// dna_generator.go — v∞ DNA traceability code generation
package lhstandard

import (
	"crypto/sha256"
	"fmt"
	"math"
	"time"
)

var tianGan = []string{"Jia","Yi","Bing","Ding","Wu","Ji","Geng","Xin","Ren","Gui"}
var diZhi = []string{"Zi","Chou","Yin","Mao","Chen","Si","Wu","Wei","Shen","You","Xu","Hai"}
var shiChen = []string{"ZiShi","ChouShi","YinShi","MaoShi","ChenShi","SiShi","WuShi","WeiShi","ShenShi","YouShi","XuShi","HaiShi"}

type Hexagram struct{ Symbol, EnName, CnName, Domain string }

var hexagrams = []Hexagram{
	{"䷀","Qian","乾","governance"},{"䷁","Kun","坤","archive"},
	{"䷂","Zhun","屯","init"},{"䷃","Meng","蒙","learn"},
	{"䷄","Xu","需","async"},{"䷅","Song","讼","legal"},
	{"䷜","Kan","坎","engine"},{"䷝","Li","离","audit"},
	{"䷲","Zhen","震","security"},{"䷳","Gen","艮","privacy"},
	{"䷸","Xun","巽","deploy"},{"䷹","Dui","兑","trust"},
	{"䷾","JiJi","既济","complete"},{"䷿","WeiJi","未济","progress"},
}

var taskHexagramMap = map[string]string{
	"default":"governance","code":"engine","deploy":"deploy","audit":"audit",
	"security":"security","archive":"archive","init":"init","learn":"learn",
	"legal":"legal","privacy":"privacy","trust":"trust","complete":"complete","progress":"progress",
}

type DNAGenerator struct {
	UID, Device, Locale string
}

func NewDNAGenerator(uid, device, locale string) *DNAGenerator {
	if uid == "" { uid = "9622" }
	if device == "" { device = "HM-9622-001" }
	if locale == "" { locale = "Asia/Shanghai" }
	return &DNAGenerator{uid, device, locale}
}

func (g *DNAGenerator) Generate(taskType, action, version string) string {
	if taskType == "" { taskType = "default" }
	if action == "" { action = "WRAP" }
	if version == "" { version = "V1.0" }
	now := time.Now().UTC()
	stem := g.computeStemBranch(now)
	hex := g.selectHexagram(taskType)
	body := fmt.Sprintf("ADAPTER-%s-%s-%s", toUpper(taskType), toUpper(action), version)
	raw := fmt.Sprintf("%s%s%s%s%s%s%s%s%s", stem["year"],stem["month"],stem["day"],stem["shichen"],hex.Symbol,hex.EnName,body,g.Device,now.Format(time.RFC3339))
	h := sha256.Sum256([]byte(raw))
	return fmt.Sprintf("#LongHun⚡️%s·%s·%s·%s·%s%s-%s-%x", stem["year"],stem["month"],stem["day"],stem["shichen"],hex.Symbol,hex.EnName,body,h[:4])
}

func toUpper(s string) string {
	r := []rune(s)
	for i, c := range r {
		if c >= 'a' && c <= 'z' { r[i] = c - 32 }
	}
	return string(r)
}

func (g *DNAGenerator) computeStemBranch(dt time.Time) map[string]string {
	cycleYear, cycleMonth := 1984, []int{2,4,6,8,10,0,2,4,6,8,10,0}
	y := dt.Year(); m := int(dt.Month()); doy := dt.YearDay(); h := dt.Hour()
	ys := (y-cycleYear) % 10; if ys < 0 { ys += 10 }
	yb := (y-cycleYear) % 12; if yb < 0 { yb += 12 }
	mb := cycleMonth[(y-cycleYear)%10]; if mb < 0 { mb += 10 }
	ms := (mb + m - 1) % 10; if ms < 0 { ms += 10 }
	mbr := (m + 1) % 12
	ds := (y - 1900 + (y-1900)/4 + doy) % 10; if ds < 0 { ds += 10 }
	db := (y - 1900 + (y-1900)/4 + doy) % 12; if db < 0 { db += 12 }
	si := h / 2
	return map[string]string{
		"year": tianGan[ys]+diZhi[yb], "month": tianGan[ms]+diZhi[mbr],
		"day": tianGan[ds]+diZhi[db], "shichen": shiChen[si],
	}
}

func (g *DNAGenerator) selectHexagram(taskType string) Hexagram {
	domain := taskHexagramMap[taskType]
	if domain == "" { domain = "governance" }
	for _, h := range hexagrams {
		if h.Domain == domain { return h }
	}
	return hexagrams[0]
}

func absMod(a, b int) int { r := a % b; if r < 0 { r += b }; return r }
var _ = absMod // suppress unused
var _ = math.Abs

File: adapters/rust/src/dna_generator.rs
use chrono::{DateTime, Utc, Timelike, Datelike};
use sha2::{Sha256, Digest};
use std::collections::HashMap;

const TIAN_GAN: [&str; 10] = ["Jia","Yi","Bing","Ding","Wu","Ji","Geng","Xin","Ren","Gui"];
const DI_ZHI: [&str; 12] = ["Zi","Chou","Yin","Mao","Chen","Si","Wu","Wei","Shen","You","Xu","Hai"];
const SHI_CHEN: [&str; 12] = ["ZiShi","ChouShi","YinShi","MaoShi","ChenShi","SiShi","WuShi","WeiShi","ShenShi","YouShi","XuShi","HaiShi"];

pub fn generate_dna(uid: &str, device: &str, task_type: &str, action: &str, version: &str) -> String {
    let now = Utc::now();
    let stem = compute_stem_branch(now);
    let hexagram = select_hexagram(task_type);
    let version = if version.is_empty() { "V1.0" } else { version };
    let body = format!("ADAPTER-{}-{}-{}", task_type.to_uppercase(), action.to_uppercase(), version);
    let raw = format!("{}{}{}{}{}{}{}{}{}", stem.year, stem.month, stem.day, stem.shichen, hexagram.0, hexagram.1, body, device, now.to_rfc3339());
    let hash = Sha256::digest(raw.as_bytes());
    let hash8 = format!("{:x}", hash);
    format!("#LongHun⚡️{}·{}·{}·{}·{}{}-{}-{}", stem.year, stem.month, stem.day, stem.shichen, hexagram.0, hexagram.1, body, &hash8[..8])
}

struct StemBranch { year: String, month: String, day: String, shichen: String }

fn compute_stem_branch(dt: DateTime<Utc>) -> StemBranch {
    let cycle_year: i32 = 1984;
    let cycle_month: [i32; 12] = [2,4,6,8,10,0,2,4,6,8,10,0];
    let y = dt.year(); let m = dt.month() as i32; let doy = dt.ordinal() as i32; let h = dt.hour() as i32;

    let ys = ((y - cycle_year) % 10 + 10) % 10;
    let yb = ((y - cycle_year) % 12 + 12) % 12;

    let mb = cycle_month[((y - cycle_year) % 10 + 10) as usize % 10];
    let ms = ((mb + m - 1) % 10 + 10) % 10;
    let mbr = ((m + 1) % 12 + 12) % 12;

    let ds = ((y - 1900 + (y - 1900) / 4 + doy) % 10 + 10) % 10;
    let db = ((y - 1900 + (y - 1900) / 4 + doy) % 12 + 12) % 12;

    let si = (h / 2) as usize;

    StemBranch {
        year: format!("{}{}", TIAN_GAN[ys as usize], DI_ZHI[yb as usize]),
        month: format!("{}{}", TIAN_GAN[ms as usize], DI_ZHI[mbr as usize]),
        day: format!("{}{}", TIAN_GAN[ds as usize], DI_ZHI[db as usize]),
        shichen: SHI_CHEN[si].to_string(),
    }
}

fn select_hexagram(task_type: &str) -> (&'static str, &'static str) {
    let domain = match task_type {
        "code" => "engine", "deploy" => "deploy", "audit" => "audit",
        "security" => "security", "archive" => "archive", "init" => "init",
        "learn" => "learn", "legal" => "legal", "privacy" => "privacy",
        "trust" => "trust", "complete" => "complete", "progress" => "progress",
        _ => "governance",
    };
    let hexagrams: [(&str,&str,&str); 14] = [
        ("䷀","Qian","governance"),("䷁","Kun","archive"),("䷂","Zhun","init"),
        ("䷃","Meng","learn"),("䷄","Xu","async"),("䷅","Song","legal"),
        ("䷜","Kan","engine"),("䷝","Li","audit"),("䷲","Zhen","security"),
        ("䷳","Gen","privacy"),("䷸","Xun","deploy"),("䷹","Dui","trust"),
        ("䷾","JiJi","complete"),("䷿","WeiJi","progress"),
    ];
    for h in hexagrams.iter() { if h.2 == domain { return (h.0, h.1) } }
    (hexagrams[0].0, hexagrams[0].1)
}
